lib/storage.ml

```diff
@@ -49,15 +49,25 @@ let epoch_end task sr vdi =
        Client.VDI.epoch_end (Xenops_task.get_dbg task) sr vdi
     ) ()
 
-let attach_and_activate task _vm dp sr vdi read_write =
+let attach_and_activate_impl ~task ~_vm ~vmdomid ~dp ~sr ~vdi ~read_write =
+  let vm = Storage_interface.Vm.of_string vmdomid in
   let result =
     Xenops_task.with_subtask task (Printf.sprintf "VDI.attach2 %s" dp)
-      (transform_exception (fun () -> Client.VDI.attach2 "attach_and_activate" dp sr vdi read_write)) in
+      (transform_exception (fun () -> Client.VDI.attach2 "attach_and_activate_impl" dp sr vdi vm read_write)) in
 
   Xenops_task.with_subtask task (Printf.sprintf "VDI.activate %s" dp)
-    (transform_exception (fun () -> Client.VDI.activate "attach_and_activate" dp sr vdi));
+    (transform_exception (fun () -> Client.VDI.activate "attach_and_activate_impl" dp sr vdi));
   result
 
+let attach_and_activate task _vm vmdomid dp sr vdi read_write =
+  let curried_attach_and_activate = attach_and_activate_impl ~task ~_vm ~dp ~sr ~vdi ~read_write in
+  match vmdomid with
```

Comment:
I'd rather make domid required and fail if we cannot determine what it is

Reply:
Fixed